File: Cpp/AdventOfCode/src/2015/d2_ElfWrappingPaper.cpp
#pragma once

#include "2015/d2_ElfWrappingPaper.h"

SOLUTION(2015, 2) {
    constexpr size_t CalculateWrappingPaper(size_t length, size_t width, size_t height)
    {
        auto d1 = height * width;
        auto d2 = height * length;
        auto d3 = width * length;

        if (d1 <= d2 && d1 <= d3) {
            return (d1 * 3) + (d2 * 2) + (d3 * 2);
        }
        else if (d2 <= d1 && d2 <= d3) {
            return (d1 * 2) + (d2 * 3) + (d3 * 2);
        }
        else {
            return (d1 * 2) + (d2 * 2) + (d3 * 3);
        }
    }

    constexpr size_t CalculateRibbon(size_t length, size_t width, size_t height)
    {
        u64 area = height * width * length;
        if (height >= width && height >= length) {
            return area + width + width + length + length;
        }
        else if (width >= height && width >= length) {
            return area + height + height + length + length;
        }
        else if (length >= width && length >= height) {
            return area + width + width + height + height;
        }
        return 0;
    }

    constexpr void ParseLine(std::string_view line, size_t& outLength, size_t& outWidth, size_t& outHeight) {
        auto s = Constexpr::Split(line, "x");
        Constexpr::ParseNumber(s[0], outLength);
        Constexpr::ParseNumber(s[1], outWidth);
        Constexpr::ParseNumber(s[2], outHeight);
    }

    constexpr std::string Solve(const auto& lines, auto func) {
        size_t result = 0;
        for (const auto& line : lines) {
            size_t length, width, height;
            ParseLine(line, length, width, height);
            result += func(length, width, height);
        }

        return Constexpr::ToString(result);
    }
    PART_ONE() {
        return Solve(Lines, CalculateWrappingPaper);
    }

    PART_TWO() {
        return Solve(Lines, CalculateRibbon);
    }

    TESTS() {
        static_assert(CalculateWrappingPaper(2, 3, 4) == 58);
        static_assert(CalculateWrappingPaper(1, 1, 10) == 43);

        static_assert(CalculateRibbon(2, 3, 4) == 34);
        static_assert(CalculateRibbon(1, 1, 10) == 14);
        static_assert(CalculateRibbon(2, 2, 2) == 16);

        return true;
    }
}

File: Rust/rust_aoc/src/year_2015/d2_elf_wrapping_paper.rs

pub fn part_one(lines: &Vec<String>) -> String {
    solve(lines, calc_paper).to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    solve(lines, calc_ribbon).to_string()
}

fn solve(lines: &Vec<String>, func: fn(u32, u32, u32)->u32) -> u32 {
    let mut result: u32 = 0;
    for line in lines {
        let parts: Vec<u32> = line.split("x").map(|d| d.parse().unwrap()).collect();
        let (width, height, depth) = (parts[0], parts[1], parts[2]);
        result += func(width, height, depth)
    }

    result
}

fn calc_paper(width: u32, height: u32, depth: u32) -> u32 {
    let d1 = width * height;
    let d2 = width * depth;
    let d3 = height * depth;

    (d1 * 2) + (d2 * 2) + (d3 * 2) + std::cmp::min(d1, std::cmp::min(d2, d3))
}

fn calc_ribbon(width: u32, height: u32, depth: u32) -> u32 {
    let area = width * height * depth;
    (area + (width * 2) + (height * 2) + (depth * 2)) - (std::cmp::max(width, std::cmp::max(height, depth)) * 2)
}